src/components/Layer/Layer.tsx

```diff
@@ -1,74 +1,133 @@
+
 import * as React from 'react';
-import * as ReactDOM from 'react-dom';
 import { ILayerProps } from './Layer.Props';
-import { LayerHost, ILayer } from './LayerHost';
-import { getId } from '../../utilities/object';
+import { Portal, IPortalProps } from '../Portal/Portal';
+import { layerPortalNexusKey } from './LayerPortalNexusKey';
+import { setVirtualParent } from '../../utilities/DomUtils';
 import './Layer.scss';
 
-const LAYER_HOST_ELEMENT_ID = 'ms-layer-host';
+export interface ILayerContext {
+  isInLayer?: boolean;
+}
+
+export const LAYER_CONTEXT_PROP_TYPES = {
+  isInLayer: React.PropTypes.bool
+};
 
-let _layerHost: LayerHost;
+const LayerPortal: (new (props: IPortalProps<void>, ...args: any[]) => React.Component<IPortalProps<void>, any>) = Portal;
 
+/**
+ * Component which projects child elements onto a layer host.
+ *
+ * @export
+ * @class Layer
+ * @extends {React.Component<ILayerProps, {}>}
+ */
 export class Layer extends React.Component<ILayerProps, {}> {
-  public static contextTypes = {
-    isInLayer: React.PropTypes.bool
-  };
+  public static contextTypes = LAYER_CONTEXT_PROP_TYPES;
 
-  public context: {
-    isInLayer: boolean;
-  };
+  public context: ILayerContext;
 
-  private _layer: ILayer;
+  private _layerElement: HTMLDivElement;
+  private _layerContentElement: HTMLDivElement;
 
-  constructor(props?: ILayerProps) {
-    super(props);
+  constructor(layerProps: ILayerProps, context: any) {
+    super(layerProps, context);
 
-    this._layer = {
-      id: getId('Layer'),
-      children: props.children
-    };
+    this._onLayerRef = this._onLayerRef.bind(this);
+    this._onLayerContentRef = this._onLayerContentRef.bind(this);
   }
 
   public render(): JSX.Element {
-    let { isInLayer } = this.context;
+    let {
+      children
+    } = this.props;
 
-    return isInLayer ? this.props.children as JSX.Element : null;
-  }
+    let {
+      isInLayer = false
+    } = this.context;
+
+    if (isInLayer) {
+      return children as JSX.Element;
```

Comment:
We may not need this hack any more with this change. if you can find all the element.contains references and use the helper you added, I think we should be good. The real test is "context menu in a callout" (